lua/plenary/async_lib2/async.lua

```diff
@@ -0,0 +1,139 @@
+  local stat = f.first(...)
+
+  if not stat then
+    error(string.format("The coroutine failed with this message: %s", f.second(...)))
+  end
+
+  if co.status(thread) == "dead" then
+    (callback or function() end)(select(2, ...))
+  else
+    local returned_function = f.second(...)
+    assert(type(returned_function) == "function", "type error :: expected func")
+    local stat, msg = pcall(returned_function, vararg.rotate(step, select(3, ...)))
+    if not stat then
+      error(('Failed to call leaf async function: %s'):format(msg))
+    end
+  end
+end
+
+---@class Future
+---Something that will give a value when run
+
+---Executes a future with a callback when it is done
+---@param async_function Future: the future to execute
+---@param callback function: the callback to call when done
+local execute = function(async_function, callback)
+  assert(type(async_function) == "function", "type error :: expected func")
+
+  local thread = co.create(async_function)
+
+  local step
+  step = function(...)
+    callback_or_next(step, thread, callback, co.resume(thread, ...))
+  end
+
+  step()
+end
+
+local add_leaf_function
+do
+  ---A table to store all leaf async functions
+  local leaf_table = setmetatable({}, {
+    __mode = "k",
+  })
+
+  add_leaf_function = function(async_func)
+    assert(leaf_table[async_func] == nil, "Async function should not already be in the table")
+    leaf_table[async_func] = true
+  end
+
+  function M.is_leaf_function(async_func)
+    return leaf_table[async_func] ~= nil
+  end
+end
+
+---Creates an async function with a callback style function.
+---@param func function: A callback style function to be converted. The last argument must be the callback.
+---@param argc number: The number of arguments of func. Must be included.
+---@return function: Returns an async function
+M.wrap = function(func, argc)
+  if type(func) ~= "function" then
+    traceback_error("type error :: expected func, got " .. type(func))
+  end
+
+  if type(argc) ~= "number" then
+    traceback_error("type error :: expected number, got " .. type(argc))
+  end
+
+  local function leaf(...)
+    local nargs = select('#', ...)
+
+    if not (nargs == argc - 1 or nargs == argc) then
+      print(('Expected %s or %s number of arguments, got %s'):format(argc - 1, argc, nargs))
+    end
+
+    if nargs == argc then
+      return func(...)
+    else
+      return co.yield(func, ...)
+    end
+  end
+
+  add_leaf_function(leaf)
+
+  return leaf
+end
+
+---Use this to either run a future concurrently and then do something else
+---or use it to run a future with a callback in a non async context
+---@param future Future
+---@param callback function
+M.run = function(async_function, callback)
+  if M.is_leaf_function(async_function) then
+    async_function(callback)
+  else
+    execute(async_function, callback)
+  end
+end
+
+---Doenst do anything, just for compat
+M.await = function(...)
+  return ...
+end
```

Comment:
we should just remove this. don't worry about compat

Reply:
Hmm, one thing I've noticed when using async is that it's good to know in the code when a particular statement may not run atomically, meaning it may execute over several iterations of the event loop.

For example:

```lua
local b = vim.api.nvim_get_current_buf()
await(some_op)

-- Just done an await so need to check buf still exists and
-- hasn't been deleted whilst the coroutine was suspended.
if vim.api.nvim_buf_is_valid(b) then
    do_buf_action(b)
end
```

Without `await`, it makes this situation a lot harder to catch.

Thoughts?

Reply:
Hmmm, but the `await` does literally nothing -- so I don't like having it around because then you expect, as the developer. that _only` things with `await(...)` are awaitable, whereas that's not the case anymore.

I do see what you're saying about running the code and then seeing that a buffer that was valid is no longer valid because something happened in the meantime. That is a difficult part to answer.

Reply:
Maybe it's just a case of adopting a naming convention for async functions and commenting code more. 🤷‍♂️

Reply:
I talked with bfredl about this also and he said just to have a naming convention and comment which I agree. The benefits to not use functions that return futures is that they are faster and are just regular functions, so you can use them as callbacks or iterators.

About needing to check if buffers still exist there is not really a good way to deal with this other than manually writing if statements to check.